src/Jellyfin.Database/Jellyfin.Database.Implementations/Locking/OptimisticLockBehavior.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+using Polly;
+
+namespace Jellyfin.Database.Implementations.Locking;
+
+/// <summary>
+/// Defines a locking mechanism that will retry any write operation for a few times.
+/// </summary>
+public class OptimisticLockBehavior : IEntityFrameworkCoreLockingBehavior
+{
+    private readonly Policy _writePolicy;
+    private readonly AsyncPolicy _writeAsyncPolicy;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="OptimisticLockBehavior"/> class.
+    /// </summary>
+    public OptimisticLockBehavior()
+    {
+        System.Collections.Generic.IEnumerable<TimeSpan> sleepDurations = [
+            TimeSpan.FromMilliseconds(50),
+            TimeSpan.FromMilliseconds(50),
+            TimeSpan.FromMilliseconds(250),
+            TimeSpan.FromMilliseconds(150),
+            TimeSpan.FromMilliseconds(500),
+            TimeSpan.FromMilliseconds(500),
+            TimeSpan.FromSeconds(3)
+        ];
```

Comment:
I'm not a fan at all of the "sleep" approach to retries on errors. If the locking operation finishes and another one enters immediately another lock, this one would be locked again once the sleep time ends, waiting yet another time. Plus we are not sure they're performed in order.

There are definitely a lot of things we can't really control properly with time-based retries. What if we:

- Introduce a lock (like in the pessimistic lock approach) in the read/write operations.
- If another operation needs to be performed when a lock is in place, do it anyway (optimistic behaviour).
    - If it doesn't fail: All right!
    - If it does fail: "Listen" to the lock change and perform as soon as the lock is released.

Reply:
they would not lock at the same time and if two writes conflict with another, one wins while the other looses so the assumption is that by the time either by chance this connection wins or its actually an issue.

Overall i am not a huge fan of it too, but its not the default option and should only be seen as experimental for users that still have issues with the two other options.

Reply:
Using an exponential back off to spread it out more, along with random jitter to avoid the Thundering herd problem could be a more robust solution here.

Reply:
for the exponential backoff, i would strongly argue against that as it will likely not contribute in any meaningfull way here but only complicate and prolong the operations, resolution of the operation in the current timeframe is still acceptable (though the 3 sec one is just a last resort matter).

For the random addition that makes sense at face value, but again does not add really. It does not really matter if an operation is +/- 10ms shifted as when another operation is blocking, that can happen at any given time and even if the colision were to happen at the __exact__ same time, still one operation always wins so next retry is fair game and the other parallel operation has already been resolved.